### ISSUE TITLE
Feature Request: Context-Preserving Middleware Chaining

### ISSUE THREAD

**Issue by rvndev** [chungweileong94/server-act#26]
Currently, `server-act` overwrites previous context when chaining multiple middlewares. It would be very useful if, instead of the current

**Comment by chungweileong94** [comment]
Thought of that before, but it wasn't too difficult to achieve that in the current state, by chaining multiple middlewares (or functions, which they simply are).  Maybe we can have a helper function like `chainMiddlewares(midlewareOne, middlewareTwo, ...)`, so it will look something like
```ts
serverAct
  .middleware(chainMiddleware(middlewareOne, middlewareTwo, ...))
  .action(...)
```

I also thought of chaining middleware like `serverAct.middleware(...).middleware(...).action(...)`, but I worry about it this will create any overhead. I might be overthinking, but feedbacks are welcome🙂

**Comment by chungweileong94** [comment]
More on the chaining middleware, it could be challenging to make it type-safe as well.

**Comment by chungweileong94** [comment]
Before we have an actual middleware chaining implementation, we will prevent people from chaining multiple duplicate methods at type-level.

**Comment by cdimitroulas** [comment]
@chungweileong94 thanks for this great package! Are you still considering to add a `chainMiddleware` helper or to add the ability to chain middleware?
Either of those would be quite helpful I think

**Comment by chungweileong94** [comment]
@cdimitroulas Yeah I still do, just that I haven't had time for this lately. But feel free to submit a PR🙂

**Comment by chungweileong94** [comment]
So I was trying to implement the chain middleware, turns out it's not an easy thing to achieve, in terms of type-safety. On top of that, I also feel like people might expect that middleware should look something like the way how Express.js middleware works, which I'm not too familiar with the implementation. Feel free to get in touch if you know a thing or two🙏

In case someone want to achieve the chain middleware in the current state, this is how you can do it:
```ts
function middlewareA() {
  return {a: "1"};
}

function middlewareB() {
  return {B: "1"};
}

const action = serverAct
  .middleware(() => {
    return {
      ...middlewareA(),
      ...middlewareB()
    };
  })
  .action(() => {});
```
Not elegant, but it should get the job done.
